ecosystem/sep-0023.md

```diff
@@ -0,0 +1,140 @@
+account IDs that are multiplexed across users or separate transactions
+and hence must be used in conjunction with a particular memo ID.
+
+## Motivation
+
+A common pattern in the Stellar ecosystem is for services to share a
+single Stellar account ID across many users or independent
+transactions, relying on the memo ID to disambiguate incoming
+payments.  Unfortunately, experience shows that people frequently
+forget to include the memo ID, resulting in either lost funds or
+onerous support calls.  Hence, we need a way to specify an address and
+memo together in a way that is difficult for end-users to misuse.
+
+## Abstract
+
+A new strkey format is introduced for _multiplexed_ addresses that can
+only be used as a destination address and only in conjunction with a
+particular memo ID.
+
+## Specification
+
+To multiplex a Stellar account over multiple users, the account must
+be used in conjunction with a particular memo ID.  Because the memo
+field is per-transaction rather than per-operation, this places
+restrictions on the use of such multiplexing.  In particular:
+
+* A multiplexed address can occur only as the `destination` field of a
+  `PAYMENT`, `PATH_PAYMENT`, or `ACCOUNT_MERGE` operation.
+
+* At most one multiplexed address can appear in a given transaction.
+
+### Native representation
+
+Because multiplexed addresses cannot be used everyplace a normal
+account can be, implementations must represent them using a distinct
+type.  Since this representation depends on the language, we give a
+suggested implementation in XDR:
+
+~~~ {.c}
+struct MultiplexedAccountID {
+  uint64 *memoID;
+  AccountID accountID;
+};
+~~~
+
+If `memoID` is `NULL`, the structure should be considered equivalent
+to a simple `AccountID`.  Otherwise, it is a multiplexed `AccountID`
+subject to the restrictions above, and must have a `MEMO_ID`.
+
+### Encoding
+
+Strkey provides a compact ASCII format for ED25519 public keys,
+ED25519 private keys (also known as seeds), pre-authorized transaction
+hashes, and hash-x signers (which provide signing authority upon
+revelation of a SHA-256 preimage). We add a fifth type representing
+multiplexed account IDs named by ED25519 keys.  Each of these types
+has a corresponding version byte, which determines the first character
+of the strkey encoding:
+
+| Key type                | Version byte | First char | Muxed |
+| ----------------------- | ------------ | ---------- | ----- |
+| STRKEY\_PUBKEY\_ED25519 | 6 << 3       | G          | no    |
+| STRKEY\_PUBKEY\_ED25519 | 12 << 3      | M          | yes   |
+| STRKEY\_SEED\_ED25519   | 18 << 3      | S          | no    |
+| STRKEY\_PRE\_AUTH\_TX   | 19 << 3      | T          | no    |
+| STRKEY\_HASH\_X         | 23 << 3      | X          | no    |
+
+
+The following steps transform a binary key into a strkey:
+
+1.  Start with the appropriate version byte from the above table.
+
+2.  If it is a multiplexed address, append an 8-byte memo ID in
```

Comment:
This proposal limits services to 8 byte memo IDs and not memos containing text. What's the motivation for that? Is it based on common usage in the ecosystem? A more flexible memo supporting other memo types might be easier to integrate with for services that identify transfers with alphanumeric identifiers and not integers.

Reply:
- Most exchanges indeed use MEMO_ID (uint64), though they tolerate getting a MEMO_STRING that contains a decimal representation of that id. 
- MEMO_STRING are intended for human entered strings so they don't really fit this use case.  
- I believe that 2 bytes (2^64 permutations) is a large enough space and I can't think of a scenario in which it's not enough. 32 byte memos were intended as a means to encode a return address in a memo. It's (a) not really used and (b) doesn't fit this use cases. 
- We can make this size variable but then keys could have different sizes which is a bit weird. 
- MEMO_STRING and MEMO_HASH can still be used, just not encoded in the account id.

Reply:
The 8 byte limit on unique identifiers might be what exchanges are comfortable with today, but IDs with larger spaces are more commonly used in software systems of scale, e.g. UUID (16 bytes), MongoID (12 bytes). It will make integration more straight forward for existing large systems if it's possible to reuse the same identifiers because in systems of scale expecting to generate IDs in a smaller space reliably can be challenging and require coordination.

Good point about MEMO_STRING not really being suited for this case, as it's too small at 28 bytes to encode a 16 byte ID as a hex string which would require 32 bytes.

>We can make this size variable but then keys could have different sizes which is a bit weird.

Strkeys can already be of different lengths?

Reply:
> Strkeys can already be of different lengths?

Current strkeys are always 32 bytes long after decoding. `M` keys will be longer what I think is fine because it's a breaking change anyway.